WalletWasabi/BranchNBound/BranchAndBound.cs

```diff
@@ -0,0 +1,169 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.BranchNBound
+{
+	public class BranchAndBound
+	{
+		private readonly Random _random = new();
+
+		public BranchAndBound(List<Money> availableCoins)
+		{
+			Count = availableCoins.Count;
+			SortedUTXOs = availableCoins.OrderByDescending(x => x.Satoshi).Select(c => c.Satoshi).ToArray();
+		}
+
+		private enum NextAction
+		{
+			AandB,
+			BandA,
+			A,
+			B,
+			Backtrack
+		}
+
+		private long[] SortedUTXOs { get; }
+		private int Count { get; }
+
+		public bool TryGetExactMatch(Money target, [NotNullWhen(true)] out List<Money> selectedCoins)
+		{
+			selectedCoins = new List<Money>();
+
+			if (SortedUTXOs.Sum() < target)
+			{
+				return false;
+			}
+
+			try
+			{
+				if (Search(target.Satoshi, out long[] solution))
+				{
+					selectedCoins = solution.Where(c => c > 0).Select(c => Money.Satoshis(c)).ToList();
+
+					if (CalcEffectiveValue(selectedCoins) == target)
```

Comment:
I think that this line is not necessary really as `Search` already verified that.